sklearn/impute.py

```diff
@@ -115,16 +137,37 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
     Notes
     -----
     Columns which only contained missing values at `fit` are discarded upon
-    `transform`.
+    `transform` if strategy is not "constant".
 
     """
-    def __init__(self, missing_values="NaN", strategy="mean",
-                 verbose=0, copy=True):
+    def __init__(self, missing_values=np.nan, strategy="mean",
+                 fill_value=None, verbose=0, copy=True):
         self.missing_values = missing_values
         self.strategy = strategy
+        self.fill_value = fill_value
         self.verbose = verbose
         self.copy = copy
 
+    def _validate_input(self, X):
+        allowed_strategies = ["mean", "median", "most_frequent", "constant"]
+        if self.strategy not in allowed_strategies:
+            raise ValueError("Can only use these strategies: {0} "
+                             " got strategy={1}".format(allowed_strategies,
+                                                        self.strategy))
+
+        if self.strategy in ("most_frequent", "constant"):
+            dtype = None
+        else:
+            dtype = FLOAT_DTYPES
+
+        if self.missing_values is not np.nan:
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+
+        return check_array(X, accept_sparse='csc', dtype=dtype,
```

Comment:
Should we really be converting a mixed-type dataframe to an array if we're imputing with strings?

Reply:
What is the alternative? (I maybe don't fully get your comment)

Reply:
To me the imputer would be applied to an homegeneous subset of columns of a mixed-typed dataframe thanks to the use of `ColumnTransformer` to do the dispatching as done in the example updated by this PR.